### ISSUE TITLE
Help needed: CDVPlugin class ... (pluginName: ...) does not exist

### ISSUE THREAD

**Issue by leshik** [ionic-team/capacitor#1014]
I'm trying to make the `cordova-ionic-phonegap-branch-deep-linking` plugin work, but have some issues. After some [trial & error](https://github.com/BranchMetrics/cordova-ionic-phonegap-branch-deep-linking/issues/528) I was able to make it compile, but it doesn't work with this error:
```
CDVPlugin class BranchSDK (pluginName: branchsdk) does not exist
```
@jcesarmobile @mlynch any idea where to look at to make it compatible with `Capacitor`?

**Comment by jcesarmobile** [comment]
I can't reproduce following your steps, after adding `#import <Cordova/CDVPlugin.h>` it compiles

But we can probably read `pod` tags to add them as dependencies, but would be better if the plugin author merged the PR you pointed where it uses Cordova way of doing it

**Comment by leshik** [comment]
@jcesarmobile I mean, I can compile it, but it doesn’t work for some reason – complains about missing class. I wonder if something might be wrong with cordova compatibility layer?

**Comment by jcesarmobile** [comment]
Sorry, I'm not familiar with the plugin and it requires registration, so I can't properly test it.

But I've checked and I can see branchsdk plugin in the registry, so not sure why it doesn't find it in your case.

**Comment by jcesarmobile** [comment]
yeah, as I said, I can see the plugin class properly registered.
If I try to call `initSession` it crash because I'm not registered, but it's the plugin call what makes it crash, so the plugin is registered and received the `initSession` method call, not sure what happens in your project

**Comment by jcesarmobile** [comment]
I could reproduce following your steps.

I see what the problem is, when the plugin has a pod dependency I put it in a different project that gets converted to a static framework. For some reason (maybe CocoaPods bug?), the class is not available from the static framework.
If it's not available, the plugins .js is not injected, that's why Branch is not available from js neither.
 
I did this because some pods were not working in non static project, and I couldn't find a way of knowing if it was going to work on the dynamic project or not, so I put all in the static one.

Knowing that it causes problems with the native classes, maybe we should forget about the static framework and just put all on the dynamic one, and if some plugin doesn't work because of that, just add it to the unsupported ones.

**Comment by robgungor** [comment]
I'm using ionic v4 I'm also receiving a runtime error from XCode:

` CDVPlugin class BranchSDK (pluginName: branchsdk) does not exist.`

Everything compiles fine but none of the branch functions work.

I've tried adding the plugin only in config.xml (as [specified in the docs](https://github.com/BranchMetrics/cordova-ionic-phonegap-branch-deep-linking#configure-app)).

I've also tried using 
```
ionic cordova plugin add branch-cordova-sdk
npm install @ionic-native/branch-io@4
```
[as specified in the Ionic docs](https://ionicframework.com/docs/native/branch-io)
and
`npm install @ionic-native/branch-io@latest`
and
`npm install @ionic-native/branch-io`

Everytime I've removed and added the ios platform.
```
ionic cordova platform remove ios
ionic cordova platform add ios
ionic cordova platform prepare ios
```
I still get this runtime error. 

I've tried using:
`import { BranchIo } from '@ionic-native/branch-io/ngx'`
with 
```
public branch: BranchIo;
...
this.branch.initSession().then(data => {
```

vs 
```
declare var Branch
...
Branch =  Branch || (<any>window).Branch;
Branch.initSession().then(data => {
```

I'm worried about manually copying files over to XCode because if I add/remove the platform in the future I'll have to keep copying over a hack.

Here's my config.xml:
```
<?xml version='1.0' encoding='utf-8'?>
<widget id="com.keepmakingscents" version="0.0.1" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <name>Making Scents</name>
    <description>Create, experience, share.</description>
    <author email="rob@robgungor.com" href="https://robgungor.com/">Rob Gungor</author>
    <content src="index.html" />
    <access origin="*" />
    <access origin="*facebook.com*" />
    <allow-intent href="http://*/*" />
    <allow-intent href="https://*/*" />
    <allow-intent href="tel:*" />
    <allow-intent href="sms:*" />
    <allow-intent href="mailto:*" />
    <allow-intent href="geo:*" />
    <preference name="ScrollEnabled" value="false" />
    <preference name="android-minSdkVersion" value="19" />
    <preference name="BackupWebStorage" value="none" />
    <preference name="SplashMaintainAspectRatio" value="true" />
    <preference name="SplashShowOnlyFirstTime" value="false" />
    <preference name="AutohideSplashScreen" value="false" />
    <preference name="SplashScreenDelay" value="30000" />
    <preference name="KeyboardResize" value="true" />
    <platform name="android">
        <allow-intent href="market:*" />
        <icon density="ldpi" src="resources/android/icon/drawable-ldpi-icon.png" />
        <icon density="mdpi" src="resources/android/icon/drawable-mdpi-icon.png" />
        <icon density="hdpi" src="resources/android/icon/drawable-hdpi-icon.png" />
        <icon density="xhdpi" src="resources/android/icon/drawable-xhdpi-icon.png" />
        <icon density="xxhdpi" src="resources/android/icon/drawable-xxhdpi-icon.png" />
        <icon density="xxxhdpi" src="resources/android/icon/drawable-xxxhdpi-icon.png" />
        <splash density="land-ldpi" src="resources/android/splash/drawable-land-ldpi-screen.png" />
        <splash density="land-mdpi" src="resources/android/splash/drawable-land-mdpi-screen.png" />
        <splash density="land-hdpi" src="resources/android/splash/drawable-land-hdpi-screen.png" />
        <splash density="land-xhdpi" src="resources/android/splash/drawable-land-xhdpi-screen.png" />
        <splash density="land-xxhdpi" src="resources/android/splash/drawable-land-xxhdpi-screen.png" />
        <splash density="land-xxxhdpi" src="resources/android/splash/drawable-land-xxxhdpi-screen.png" />
        <splash density="port-ldpi" src="resources/android/splash/drawable-port-ldpi-screen.png" />
        <splash density="port-mdpi" src="resources/android/splash/drawable-port-mdpi-screen.png" />
        <splash density="port-hdpi" src="resources/android/splash/drawable-port-hdpi-screen.png" />
        <splash density="port-xhdpi" src="resources/android/splash/drawable-port-xhdpi-screen.png" />
        <splash density="port-xxhdpi" src="resources/android/splash/drawable-port-xxhdpi-screen.png" />
        <splash density="port-xxxhdpi" src="resources/android/splash/drawable-port-xxxhdpi-screen.png" />
    </platform>
    <platform name="ios">
        <allow-intent href="itms:*" />
        <allow-intent href="itms-apps:*" />
        <icon height="57" src="resources/ios/icon/icon.png" width="57" />
        <icon height="114" src="resources/ios/icon/icon@2x.png" width="114" />
        <icon height="40" src="resources/ios/icon/icon-40.png" width="40" />
        <icon height="80" src="resources/ios/icon/icon-40@2x.png" width="80" />
        <icon height="120" src="resources/ios/icon/icon-40@3x.png" width="120" />
        <icon height="50" src="resources/ios/icon/icon-50.png" width="50" />
        <icon height="100" src="resources/ios/icon/icon-50@2x.png" width="100" />
        <icon height="60" src="resources/ios/icon/icon-60.png" width="60" />
        <icon height="120" src="resources/ios/icon/icon-60@2x.png" width="120" />
        <icon height="180" src="resources/ios/icon/icon-60@3x.png" width="180" />
        <icon height="72" src="resources/ios/icon/icon-72.png" width="72" />
        <icon height="144" src="resources/ios/icon/icon-72@2x.png" width="144" />
        <icon height="76" src="resources/ios/icon/icon-76.png" width="76" />
        <icon height="152" src="resources/ios/icon/icon-76@2x.png" width="152" />
        <icon height="167" src="resources/ios/icon/icon-83.5@2x.png" width="167" />
        <icon height="29" src="resources/ios/icon/icon-small.png" width="29" />
        <icon height="58" src="resources/ios/icon/icon-small@2x.png" width="58" />
        <icon height="87" src="resources/ios/icon/icon-small@3x.png" width="87" />
        <icon height="1024" src="resources/ios/icon/icon-1024.png" width="1024" />
        <splash height="1136" src="resources/ios/splash/Default-568h@2x~iphone.png" width="640" />
        <splash height="1334" src="resources/ios/splash/Default-667h.png" width="750" />
        <splash height="2208" src="resources/ios/splash/Default-736h.png" width="1242" />
        <splash height="1242" src="resources/ios/splash/Default-Landscape-736h.png" width="2208" />
        <splash height="1536" src="resources/ios/splash/Default-Landscape@2x~ipad.png" width="2048" />
        <splash height="2048" src="resources/ios/splash/Default-Landscape@~ipadpro.png" width="2732" />
        <splash height="768" src="resources/ios/splash/Default-Landscape~ipad.png" width="1024" />
        <splash height="2048" src="resources/ios/splash/Default-Portrait@2x~ipad.png" width="1536" />
        <splash height="2732" src="resources/ios/splash/Default-Portrait@~ipadpro.png" width="2048" />
        <splash height="1024" src="resources/ios/splash/Default-Portrait~ipad.png" width="768" />
        <splash height="960" src="resources/ios/splash/Default@2x~iphone.png" width="640" />
        <splash height="480" src="resources/ios/splash/Default~iphone.png" width="320" />
        <splash height="2732" src="resources/ios/splash/Default@2x~universal~anyany.png" width="2732" />
    </platform>
    <plugin name="cordova-plugin-whitelist" spec="1.3.3" />
    <plugin name="cordova-plugin-facebook4" spec="4.1.0">
        <variable name="APP_ID" value="XXX" />
        <variable name="APP_NAME" value="makingScents" />
    </plugin>
    <branch-config>
        <branch-key value="key_live_XXX" />
        <uri-scheme value="makingscents" />
        <link-domain value="makingscents.app.link" />
        <ios-team-release value="XXX" />
    </branch-config>
    <plugin name="cordova-plugin-ionic-webview" spec="^2.3.2" />
    <plugin name="cordova-plugin-statusbar" spec="2.4.2" />
    <plugin name="cordova-plugin-taptic-engine" spec="2.1.0" />
    <plugin name="cordova-plugin-ionic-keyboard" spec="2.1.3" />
    <plugin name="cordova-plugin-splashscreen" spec="5.0.2" />
    <preference name="SplashScreen" value="screen" />
    <plugin name="branch-cordova-sdk" spec="3.1.5" />
    <engine name="ios" spec="4.5.5" />
</widget>
```

Hellllllp!

**Comment by thanhnguyen92** [comment]
> 
> 
> I'm using ionic v4 I'm also receiving a runtime error from XCode:
> 
> ```
> CDVPlugin class BranchSDK (pluginName: branchsdk) does not exist.
> CDVPlugin class BranchSDK (pluginName: BranchSDK) does not exist.
> 2019-02-05 14:34:49.603140-0600 Making Scents[33499:8488466] ERROR: Plugin 'BranchSDK' not found, or is not a CDVPlugin. Check your plugin mapping in config.xml.
> 2019-02-05 14:34:49.603174-0600 Making Scents[33499:8488466] FAILED pluginJSON = ["BranchSDK356496177","BranchSDK","createBranchUniversalObject",[{"contentDescription":"Checkout my creation yes on Making Scents","title":"I made scents...","canonicalIdentifier":"content\/123","canonicalUrl":"https:\/\/example.com\/content\/123","contentIndexingMode":"private","contentMetadata":{"formulaPath":"formulae\/ZLgKTone4KPkqw4nw78P\/revisions\/YcgGeEvvnr1fJTwMlQrU"}}]]
> ```
> 
> Everything compiles fine but none of the branch functions work.
> 
> I've tried adding the plugin only in config.xml (as [specified in the docs](https://github.com/BranchMetrics/cordova-ionic-phonegap-branch-deep-linking#configure-app)).
> 
> I've also tried using
> 
> ```
> ionic cordova plugin add branch-cordova-sdk
> npm install @ionic-native/branch-io@4
> ```
> 
> [as specified in the Ionic docs](https://ionicframework.com/docs/native/branch-io)
> and
> `npm install @ionic-native/branch-io@latest`
> and
> `npm install @ionic-native/branch-io`
> 
> Everytime I've removed and added the ios platform.
> 
> ```
> ionic cordova platform remove ios
> ionic cordova platform add ios
> ionic cordova platform prepare ios
> ```
> 
> I still get this runtime error.
> 
> I've tried using:
> `import { BranchIo } from '@ionic-native/branch-io/ngx'`
> with
> 
> ```
> public branch: BranchIo;
> ...
> this.branch.initSession().then(data => {
> ```
> 
> vs
> 
> ```
> declare var Branch
> ...
> Branch =  Branch || (<any>window).Branch;
> Branch.initSession().then(data => {
> ```
> 
> I'm worried about manually copying files over to XCode because if I add/remove the platform in the future I'll have to keep copying over a hack.
> 
> Here's my config.xml:
> 
> ```
> <?xml version='1.0' encoding='utf-8'?>
> <widget id="com.keepmakingscents" version="0.0.1" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
>     <name>Making Scents</name>
>     <description>Create, experience, share.</description>
>     <author email="rob@robgungor.com" href="https://robgungor.com/">Rob Gungor</author>
>     <content src="index.html" />
>     <access origin="*" />
>     <access origin="*facebook.com*" />
>     <allow-intent href="http://*/*" />
>     <allow-intent href="https://*/*" />
>     <allow-intent href="tel:*" />
>     <allow-intent href="sms:*" />
>     <allow-intent href="mailto:*" />
>     <allow-intent href="geo:*" />
>     <preference name="ScrollEnabled" value="false" />
>     <preference name="android-minSdkVersion" value="19" />
>     <preference name="BackupWebStorage" value="none" />
>     <preference name="SplashMaintainAspectRatio" value="true" />
>     <preference name="SplashShowOnlyFirstTime" value="false" />
>     <preference name="AutohideSplashScreen" value="false" />
>     <preference name="SplashScreenDelay" value="30000" />
>     <preference name="KeyboardResize" value="true" />
>     <platform name="android">
>         <allow-intent href="market:*" />
>         <icon density="ldpi" src="resources/android/icon/drawable-ldpi-icon.png" />
>         <icon density="mdpi" src="resources/android/icon/drawable-mdpi-icon.png" />
>         <icon density="hdpi" src="resources/android/icon/drawable-hdpi-icon.png" />
>         <icon density="xhdpi" src="resources/android/icon/drawable-xhdpi-icon.png" />
>         <icon density="xxhdpi" src="resources/android/icon/drawable-xxhdpi-icon.png" />
>         <icon density="xxxhdpi" src="resources/android/icon/drawable-xxxhdpi-icon.png" />
>         <splash density="land-ldpi" src="resources/android/splash/drawable-land-ldpi-screen.png" />
>         <splash density="land-mdpi" src="resources/android/splash/drawable-land-mdpi-screen.png" />
>         <splash density="land-hdpi" src="resources/android/splash/drawable-land-hdpi-screen.png" />
>         <splash density="land-xhdpi" src="resources/android/splash/drawable-land-xhdpi-screen.png" />
>         <splash density="land-xxhdpi" src="resources/android/splash/drawable-land-xxhdpi-screen.png" />
>         <splash density="land-xxxhdpi" src="resources/android/splash/drawable-land-xxxhdpi-screen.png" />
>         <splash density="port-ldpi" src="resources/android/splash/drawable-port-ldpi-screen.png" />
>         <splash density="port-mdpi" src="resources/android/splash/drawable-port-mdpi-screen.png" />
>         <splash density="port-hdpi" src="resources/android/splash/drawable-port-hdpi-screen.png" />
>         <splash density="port-xhdpi" src="resources/android/splash/drawable-port-xhdpi-screen.png" />
>         <splash density="port-xxhdpi" src="resources/android/splash/drawable-port-xxhdpi-screen.png" />
>         <splash density="port-xxxhdpi" src="resources/android/splash/drawable-port-xxxhdpi-screen.png" />
>     </platform>
>     <platform name="ios">
>         <allow-intent href="itms:*" />
>         <allow-intent href="itms-apps:*" />
>         <icon height="57" src="resources/ios/icon/icon.png" width="57" />
>         <icon height="114" src="resources/ios/icon/icon@2x.png" width="114" />
>         <icon height="40" src="resources/ios/icon/icon-40.png" width="40" />
>         <icon height="80" src="resources/ios/icon/icon-40@2x.png" width="80" />
>         <icon height="120" src="resources/ios/icon/icon-40@3x.png" width="120" />
>         <icon height="50" src="resources/ios/icon/icon-50.png" width="50" />
>         <icon height="100" src="resources/ios/icon/icon-50@2x.png" width="100" />
>         <icon height="60" src="resources/ios/icon/icon-60.png" width="60" />
>         <icon height="120" src="resources/ios/icon/icon-60@2x.png" width="120" />
>         <icon height="180" src="resources/ios/icon/icon-60@3x.png" width="180" />
>         <icon height="72" src="resources/ios/icon/icon-72.png" width="72" />
>         <icon height="144" src="resources/ios/icon/icon-72@2x.png" width="144" />
>         <icon height="76" src="resources/ios/icon/icon-76.png" width="76" />
>         <icon height="152" src="resources/ios/icon/icon-76@2x.png" width="152" />
>         <icon height="167" src="resources/ios/icon/icon-83.5@2x.png" width="167" />
>         <icon height="29" src="resources/ios/icon/icon-small.png" width="29" />
>         <icon height="58" src="resources/ios/icon/icon-small@2x.png" width="58" />
>         <icon height="87" src="resources/ios/icon/icon-small@3x.png" width="87" />
>         <icon height="1024" src="resources/ios/icon/icon-1024.png" width="1024" />
>         <splash height="1136" src="resources/ios/splash/Default-568h@2x~iphone.png" width="640" />
>         <splash height="1334" src="resources/ios/splash/Default-667h.png" width="750" />
>         <splash height="2208" src="resources/ios/splash/Default-736h.png" width="1242" />
>         <splash height="1242" src="resources/ios/splash/Default-Landscape-736h.png" width="2208" />
>         <splash height="1536" src="resources/ios/splash/Default-Landscape@2x~ipad.png" width="2048" />
>         <splash height="2048" src="resources/ios/splash/Default-Landscape@~ipadpro.png" width="2732" />
>         <splash height="768" src="resources/ios/splash/Default-Landscape~ipad.png" width="1024" />
>         <splash height="2048" src="resources/ios/splash/Default-Portrait@2x~ipad.png" width="1536" />
>         <splash height="2732" src="resources/ios/splash/Default-Portrait@~ipadpro.png" width="2048" />
>         <splash height="1024" src="resources/ios/splash/Default-Portrait~ipad.png" width="768" />
>         <splash height="960" src="resources/ios/splash/Default@2x~iphone.png" width="640" />
>         <splash height="480" src="resources/ios/splash/Default~iphone.png" width="320" />
>         <splash height="2732" src="resources/ios/splash/Default@2x~universal~anyany.png" width="2732" />
>     </platform>
>     <plugin name="cordova-plugin-whitelist" spec="1.3.3" />
>     <plugin name="cordova-plugin-facebook4" spec="4.1.0">
>         <variable name="APP_ID" value="XXX" />
>         <variable name="APP_NAME" value="makingScents" />
>     </plugin>
>     <branch-config>
>         <branch-key value="key_live_XXX" />
>         <uri-scheme value="makingscents" />
>         <link-domain value="makingscents.app.link" />
>         <ios-team-release value="XXX" />
>     </branch-config>
>     <plugin name="cordova-plugin-ionic-webview" spec="^2.3.2" />
>     <plugin name="cordova-plugin-statusbar" spec="2.4.2" />
>     <plugin name="cordova-plugin-taptic-engine" spec="2.1.0" />
>     <plugin name="cordova-plugin-ionic-keyboard" spec="2.1.3" />
>     <plugin name="cordova-plugin-splashscreen" spec="5.0.2" />
>     <preference name="SplashScreen" value="screen" />
>     <plugin name="branch-cordova-sdk" spec="3.1.5" />
>     <engine name="ios" spec="4.5.5" />
> </widget>
> ```
> 
> Hellllllp!

I also met an exact same issue like that with Twitter Connect on Ionic v4 (https://ionicframework.com/docs/native/twitter-connect).

**Comment by Taylorsuk** [comment]
Hi Team, is there any update on Branch compatibility? I am getting the following on an iOS build. 
```
...AppDelegate+BranchSdk.m:24:97: Declaration of 'CDVPluginHandleOpenURLNotification' must be imported from module 'Cordova.CDVPlugin' before it is required
```

**Comment by jsheetzati** [comment]
I am seeing a similar message using cordova-plugin-strpnt. `CDVPlugin class StarPRNT (pluginName: starprnt) does not exist.`. 

Is there a way to determine why Capacitor is unable to use certain plugins? The project builds fine but CDVPluginManager is unable to find the plugin in `getCommandInstance`. It is difficult knowing if this is a problem with my Xcode setup or a problem with a particular Cordova plugin. Are there any troubleshooting steps when Capacitor cannot find a Cordova plugin? It would be helpful getting a guide to troubleshoot these problem plugins.

**Comment by jcesarmobile** [comment]
report plugin issues independently so they can be tracked and looked into

**Comment by codinronan** [comment]
Hi guys just following up on this. We've been asked by the Branch team to look into this issue and indeed I ran into the same problem. I have the fix working locally with an updated version of the Branch cordova plugin (yes using Cordova!) so we will be pushing this shortly to release a new version.

Probably longer term we'll move to a proper Capacitor plugin but for now the Cordova one is working.

By the way, I'm not 100% certain of this but almost that sure, the reason I can get it to work now is due to CocaoPods 1.8 -- several changes in how that version manages dependencies allow files to resolve that couldn't before.

Anyway go ahead and follow this over at the branch repository. https://github.com/BranchMetrics/cordova-ionic-phonegap-branch-deep-linking-attribution/pull/613

**Comment by jcesarmobile** [comment]
Yeah, I confirm the "CDVPlugin class (pluginName: ) does not exist" problem gets fixed when using CocoaPods 1.8.x

Glad to hear you have the plugin working on Capacitor.

Going to close since the original problem was a CocoaPods problem and the plugin will be compatible soon.

Also, I've seen there is an unofficial Capacitor plugin https://github.com/boundstate/capacitor-branch-deep-links

**Comment by dkk** [comment]
FYI: I'm on CocoaPods 1.8.4 and got `CDVPlugin class CDVUIInAppBrowser (pluginName: uiinappbrowser) does not exist.`. Generatic my project with KapselSDK3.2SP01PL2.
